### PR TITLE
Implement State DB Methods

### PR DESCRIPTION
**Pull request by terencechain**
This is part of #3064 

This PR implements the state DB methods as part of the `Database` interface accordingly

### CODE REVIEW

**Comment by prestonvanloon** [comment]
Build failing

File: beacon-chain/db/kv/state.go [hunk 1]
```diff
@@ -3,24 +3,70 @@ package kv
...
 }
 
-// HeadState returns the latest canonical state in eth2.
-// TODO(#3164): Implement.
+// HeadState returns the latest canonical state in beacon chain.
 func (k *Store) HeadState(ctx context.Context) (*pb.BeaconState, error) {
-	return nil, nil
+	var s *pb.BeaconState
+	err := k.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(blocksBucket)
```
Comment (prestonvanloon) [inline beacon-chain/db/kv/state.go]: Block bucket?

File: beacon-chain/db/kv/state.go [hunk 2]
```diff
@@ -3,24 +3,70 @@ package kv
...
+		headBlkRoot := bucket.Get(headBlockRootKey)
```
Comment (prestonvanloon) [inline beacon-chain/db/kv/state.go]: Block root? Shouldn't these be state?

File: beacon-chain/db/kv/state.go (hunk 1 above)
Comment (nisdas) [reply beacon-chain/db/kv/state.go]: +1

File: beacon-chain/db/kv/state.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/db/kv/state.go]: +1

File: beacon-chain/db/kv/state.go (hunk 1 above)
Comment (terencechain) [reply beacon-chain/db/kv/state.go]: yes, accessing `blocksBucket` to get the head block root

File: beacon-chain/db/kv/state.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/db/kv/state.go]: This function is to retrieve the head state which will be retrieved using head block root

Comment (terencechain) [reply beacon-chain/db/kv/state.go]: see L38

File: beacon-chain/db/kv/state.go [hunk 3]
```diff
@@ -3,24 +3,73 @@ package kv
...
+	"github.com/pkg/errors"
 	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
 )
 
-// State retrieval by filter criteria.
-// TODO(#3164): Implement.
-func (k *Store) State(ctx context.Context, f *filters.QueryFilter) (*pb.BeaconState, error) {
-	return nil, nil
+// State returns the saved state using input block root.
+// here the input block root was used to generate the state.
```
Comment (prestonvanloon) [inline beacon-chain/db/kv/state.go]: Punctuation. Please define the block root. Is it the signing root?

File: beacon-chain/db/kv/state.go [hunk 4]
```diff
@@ -3,24 +3,73 @@ package kv
...
+	"github.com/pkg/errors"
 	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
 )
 
-// State retrieval by filter criteria.
-// TODO(#3164): Implement.
-func (k *Store) State(ctx context.Context, f *filters.QueryFilter) (*pb.BeaconState, error) {
-	return nil, nil
+// State returns the saved state using block's signing root.
+// this particular block was used to generate the state.
```
Comment (prestonvanloon) [inline beacon-chain/db/kv/state.go]: Punctuation

**Pull request merged**